tokio/src/sync/notify.rs

```diff
@@ -913,32 +964,45 @@ impl Drop for Notified<'_> {
-            // being the only `LinkedList` available to the type.
-            unsafe { waiters.remove(NonNull::new_unchecked(waiter.get())) };
+            // Safety: called while locked
+            let w = unsafe { &mut *waiter.get() };
 
-            if waiters.is_empty() && get_state(notify_state) == WAITING {
-                notify_state = set_state(notify_state, EMPTY);
-                notify.state.store(notify_state, SeqCst);
-            }
+            if let Some(mut notify_waiters_list_ptr) = w.notify_waiters_queue.take() {
+                // We hold the `waiters` lock.
+                let notify_waiters_list = unsafe { notify_waiters_list_ptr.as_mut() };
 
-            // See if the node was notified but not received. In this case, if
-            // the notification was triggered via `notify_one`, it must be sent
-            // to the next waiter.
-            //
-            // Safety: with the entry removed from the linked list, there can be
-            // no concurrent access to the entry
-            if matches!(
-                unsafe { (*waiter.get()).notified },
-                Some(NotificationType::OneWaiter)
-            ) {
-                if let Some(waker) = notify_locked(&mut waiters, &notify.state, notify_state) {
-                    drop(waiters);
-                    waker.wake();
+                // safety: the waiter *MUST* be stored in the `notify_waiters_list`
+                // because it had a pointer to it.
+                unsafe { notify_waiters_list.remove(NonNull::new_unchecked(w)) };
+            } else {
+                let mut notify_state = notify.state.load(SeqCst);
+
+                // remove the entry from the list (if not already removed)
+                //
+                // safety: the waiter must be stored in `waiters` because it does
+                // not have pointer to any other linked list.
+                unsafe { waiters.remove(NonNull::new_unchecked(w)) };
+
+                if waiters.is_empty() && get_state(notify_state) == WAITING {
+                    notify_state = set_state(notify_state, EMPTY);
+                    notify.state.store(notify_state, SeqCst);
+                }
+
+                // See if the node was notified but not received. In this case, if
+                // the notification was triggered via `notify_one`, it must be sent
+                // to the next waiter.
+                //
+                // Safety: with the entry removed from the linked list, there can be
+                // no concurrent access to the entry
+                if matches!(
+                    unsafe { (*waiter.get()).notification },
+                    Some(NotificationType::OneWaiter)
+                ) {
+                    if let Some(waker) = notify_locked(&mut waiters, &notify.state, notify_state) {
+                        drop(waiters);
+                        waker.wake();
+                    }
                 }
```

Comment:
This can also cause weird things ... I know this was already here, and maybe the answer is to just not change it.

Imagine this sequence of actions:

 1. Create two `Notified` futures and poll them.
 2. Call `notify_one` twice.
 3. Create two new `Notified` futures.
 4. Drop the original two `Notified` futures.
 5. Both of the futures from step 3 complete.

This is weird since the futures that completed were created after the `notify_one` calls, and the fact that there are two means that we can't just pretend that this is due to the permit being stored.

Reply:
I thought about this for a while and I think that's a tough problem. The issue is that we would like to think about cancelling a `Notified` future as simply removing it from the queue, like it's written in the docs. In reality, a future is removed much earlier and at drop we only simulate what *would* happen, if the future was removed *before* receiving the permit. In this code, it is done by passing the permit to another waiting future.

However, we can pass the permit from a future A to a waiting future B if and only if at any point between receiving the permit by A and pushing B to the queue the state of `Notify` *didn't* reach `NOTIFIED`. We could check for that by enumerating all events of calling `notify_one` and enabling `Notified` futures and keeping the last number when the state was `NOTIFIED`. In `Drop` it would suffice to check if the future being dropped isn't older than this number.

Unfortunately, the problem is deeper than that, and your example demonstrates this. We can drop multiple futures, and simulating that one of them was removed from the queue before receiving the permit will affect how we simulate the same for others.

I like to think about it the following way: let $e_t$ represent the $t$-th event. $e_t = 1$ if the $t$-th event is enabling a `Notified` future and $e_t = -1$ if it is a `notify_one` call. Then we can define the number of waiters after the $t$-th event $w_t = \max(-1, w_{t-1} + e_t)$. There is a special case $w_t = -1$ if there are no waiters and the state is `NOTIFIED`. To check whether we can transfer the permit from a dropped future, we should track the last number when the state was `NOTIFIED` which is the greatest $T$ which fulfills $w_T = -1$.

The issue is that dropping a future enabled in the $t$-th event means we change $e_t$ from $1$ to $0$, what affects all newer values $w_t, w_{t+1}, \ldots$. In particular, it can change one of them to $-1$. Such a change can increase $T$ and we have to somehow account for that inside the `Drop` impl.

To sum up, we would need a complicated data structure to effectively track whether we should transfer the permit from a dropped future, for example a deque or some sort of segment tree, which probably is not feasible here. I think the best we can do is the heuristic I've mentioned earlier, but it doesn't even fix your case, so in my opinion we should probably leave this code as it is.

Reply:
I believe that we ran into a similar problem during a discussion about adding a condition variable to Tokio. To me, it seems that the best option is to just document that dropping a future that has consumed a permit from a `notify_one` call will pass on its permit as-if `notify_one` was called in the destructor.

Reply:
Sure, I will be happy to open a PR improving the docs.